pack/nodejs/helm.ts

```diff
@@ -0,0 +1,715 @@
+import * as shell from "shell-quote";
+import * as tmp from "tmp";
+import * as yaml from "js-yaml";
+
+export namespace v2 {
+    export interface ChartOpts {
+        repo: string;
+        chart: string;
+        version: string;
+
+        values?: any;
+        transformations?: transform.Transform[];
+        fetchOpts?: FetchOpts;
+    }
+
+    // A CustomResource representing a Helm Chart.
+    //
+    // `Chart` does not use Tiller. The Chart specified is copied and expanded locally; any values
+    // that would be retrieved in-cluster would be assigned fake values, and none of Tiller's
+    // server-side validity testing is executed.
+    //
+    // Important caveat: Charts often assign hard-coded names for objects; since Pulumi is not able
+    // to assign such names, any time a resource must be replaced, it will be marked as
+    // delete-before-replace. For example, in a normal Pulumi program, a change in a ConfigMap's
+    // would result in: (1) a new ConfigMap being created, (2) all Deployments referencing that
+    // ConfigMap being updated to refer to the new ConfigMap, and (3) deleting the old ConfigMap
+    // only after all the rollouts of those Deployments have succeeded. In contrast, since the
+    // chart asks for a specific name for this ConfigMap, it must be deleted before being replaced
+    // with a ConfigMap of the same name.
+    export class Chart extends pulumi.ComponentResource {
+        public readonly resources: pulumi.CustomResource[];
+
+        constructor(releaseName: string, config: ChartOpts, opts?: pulumi.ResourceOptions) {
+            super("kubernetes:helm.sh/v2:Chart", releaseName, config, opts);
+
+            // Create temporary directories and files to hold chart data and override values.
+            const overrides = tmp.fileSync({postfix: ".yaml"});
+            const chartDir = tmp.dirSync({unsafeCleanup: true});
+
+            try {
+                // Fetch chart.
+                fetch(`${config.repo}/${config.chart}`,
+                    {destination: chartDir.name, version: config.version});
+
+                // Create temporary overrides file.
+                const overrides = tmp.fileSync({postfix: ".yaml"});
+                const data = JSON.stringify(config.values || {}, undefined, "  ");
+                fs.writeFileSync(overrides.name, data);
+
+                // Does not require Tiller. From the `helm template` documentation:
+                //
+                // >  Render chart templates locally and display the output.
+                // >
+                // > This does not require Tiller. However, any values that would normally be
+                // > looked up or retrieved in-cluster will be faked locally. Additionally, none
+                // > of the server-side testing of chart validity (e.g. whether an API is supported)
+                // > is done.
+                const chart = `${chartDir.name}/${config.chart}`;
```

Comment:
Are these names guaranteed to be shell-safe - or should these also be `shell.quote`d?

Reply:
Names are supposed to be DNS-1123 compliant, but on second thought they should be escaped because people can put bad things here.